File: algorithms/[547]省份数量/solution.py
from typing import List


class UnionFind:
    def __init__(self, n: int):
        self.parent = [i for i in range(n)]
        self.rank = [1] * n

    def union(self, i: int, j: int) -> None:
        x, y = self.find(i), self.find(j)

        if self.rank[x] <= self.rank[y]:
            self.parent[x] = y
        else:
            self.parent[y] = x

        # 如果深度相同且根节点不同，则新的根节点的深度+1
        if self.rank[x] == self.rank[y] and x is not y:
            self.rank[y] += 1

    def find(self, x: int) -> int:
        if self.parent[x] is not x:
            self.parent[x] = self.find(self.parent[x])

        return self.parent[x]


class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        uf = UnionFind(n)

        for i in range(n):
            for j in range(i + 1, n):
                if isConnected[i][j] == 1:
                    uf.union(i, j)

        res = [val for idx, val in enumerate(uf.parent) if idx == val]
        return len(res)

File: algorithms/[547]省份数量/solution.ts
// topics = ["图", "深度优先搜索", "广度优先搜索", "并查集"]

// 已知无向图的邻接矩阵，求连通分量
function findCircleNum(isConnected: number[][]): number {
  // dfs
  const n = isConnected.length;
  const visited = new Set<number>();
  let res: number = 0;

  const dfs = function (i: number): void {
    visited.add(i);
    for (let j = 0; j < n; j += 1) {
      if (isConnected[i][j] === 1 && !visited.has(j)) {
        dfs(j);
      }
    }
  };

  for (let i = 0; i < n; i += 1) {
    if (!visited.has(i)) {
      dfs(i);
      res += 1;
    }
  }

  return res;
}

function findCircleNum2(isConnected: number[][]): number {
  // bfs
  const n = isConnected.length;
  const visited = new Set<number>();
  let queue: number[] = [];
  let res: number = 0;

  for (let i = 0; i < n; i += 1) {
    if (visited.has(i)) continue;

    queue.push(i);
    while (queue.length) {
      const cur = queue.shift()!;
      visited.add(cur);
      for (let j = 0; j < n; j += 1) {
        if (isConnected[cur][j] === 1 && !visited.has(j)) {
          queue.push(j);
        }
      }
    }

    res += 1;
  }

  return res;
}

function findCircleNum3(isConnected: number[][]): number {
  // Union-Find
  const n = isConnected.length;
  const parent = Array.from({ length: n }, (_, i) => i);

  const union = function (i: number, j: number): void {
    parent[find(i)] = find(j);
  };

  const find = function (x: number): number {
    if (parent[x] === x) {
      return x;
    }

    // return find(parent[x]);

    // 优化(路径压缩)：把沿途的每个节点的父节点都设为根节点
    parent[x] = find(parent[x]);
    return parent[x];
  };

  for (let i = 0; i < n; i += 1) {
    for (let j = i + 1; j < n; j += 1) {
      if (isConnected[i][j] === 1) {
        union(i, j);
      }
    }
  }

  return parent.filter((ele, index) => ele === index).length;
}

function findCircleNum4(isConnected: number[][]): number {
  // Union-Find & 按秩合并
  // 参考 https://zhuanlan.zhihu.com/p/93647900
  const n = isConnected.length;
  const parent = Array.from({ length: n }, (_, i) => i);
  const rank = Array.from({ length: n }, () => 1);

  const union = function (i: number, j: number): void {
    const x = find(i);
    const y = find(j);

    if (rank[x] <= rank[y]) {
      parent[x] = y;
    } else {
      parent[y] = x;
    }

    if (rank[x] === rank[y] && x !== y) {
      rank[y] += 1;
    }
  };

  const find = function (x: number): number {
    if (parent[x] !== x) {
      parent[x] = find(parent[x]);
    }

    return parent[x];
  };

  for (let i = 0; i < n; i += 1) {
    for (let j = i + 1; j < n; j += 1) {
      if (isConnected[i][j] === 1) {
        union(i, j);
      }
    }
  }

  return parent.filter((ele, index) => ele === index).length;
}
